### PR TITLE
add addTypesToCompiler and addTypesToInterperter method for ProtoDef

### PR DESCRIPTION
**Pull request by extremeheat**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -85,6 +84,12 @@ Provide the big-endian protodef instance used to parse and serialize nbt.
 
 Provide the little-endian protodef instance used to parse and serialize little endian nbt.
 
+### addTypesToCompiler (type, compiler)
+Adds prismarine-nbt types to an ProtoDef compiler instance
+
+### addTypesToInterperter (type, interperter)
```
Comment (rom1504) [inline README.md:90]: typo

File: nbt.js [hunk 2]
```diff
@@ -6,18 +6,37 @@ const beNbtJson = JSON.stringify(require('./nbt.json'))
...
   compiler.addTypesToCompile(JSON.parse(proto))
+}
+
+function addTypesToInterperter (type, compiler) {
+  compiler.addTypes(require('./compound'))
+  compiler.addType('nbtTagName', 'shortString')
+  compiler.addTypes(require('./compiler-optional').interpert)
+  let proto = beNbtJson
+  if (type === 'littleVarint') {
+    compiler.addTypes(require('./compiler-zigzag').interpert)
```
Comment (rom1504) [inline nbt.js:29]: interpret

File: compiler-optional.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
+function readOptionalNbt (buffer, offset, { tagType } = { tagType: 'nbt' }, rootNode) {
```
Comment (rom1504) [inline compiler-optional.js:1]: let's remove compiler from the name since it provides both version / and let's do the same for other files

**Comment by rom1504** [comment]
nice, would be good to add a simple test using this .addTypesToCompiler and using 'nbt' in a custom protodef proto
so we don't have to rely on nmp tests to make sure this code works

**Comment by rom1504** [comment]
But also let's try to integrate this in nmp PR and see if that works well

File: nbt.json [hunk 4]
```diff
@@ -121,8 +135,25 @@
...
+    {
+      "tagType": "anonymousNbt"
+    }
+  ],
+  "optionalNbt": [
+    "optionalNbtType",
+    {
+      "tagType": "nbt"
+    }
   ]
```
Comment (extremeheat) [inline nbt.json:158]: These don't have to be here, but can be put into all of the the protocol.json's if needed. `optionalNbtType` is the base type that can take a custom `tagType` to read from the stream. I kept `optionalNbt` naming here to refer to the default `nbt` tag just for compatibility with current protocol.json's

File: bench/compiled_nbt.js [hunk 5]
```diff
@@ -3,22 +3,19 @@ const { performance } = require('perf_hooks')
...
-  if (error) {
-    throw error
-  }
-
-  const proto = new ProtoDef()
-  proto.addTypes(require('../compound'))
-  proto.addTypes(require('../nbt.json'))
+function main (nbTests = 10000) {
+  const buffer = fs.readFileSync(__dirname + '/../sample/bigtest.nbt') // eslint-disable-line n/no-path-concat
+  const validate = false
```
Comment (extremeheat) [inline bench/compiled_nbt.js:12]: This fails with validation on with protodef-validator
```json
{
  "keyword": "enum",
  "dataPath": "",
  "schemaPath": "#/oneOf/0/enum",
  "params": {
    "allowedValues": [
      "native"
    ]
  },
  "message": "should be equal to one of the allowed values",
  "schema": [
    "native"
  ],
  "parentSchema": {
    "enum": [
      "native"
    ]
  },
  "data": [
    "container",
    [
      {
        "name": "type",
        "type": "nbtMapper"
      },
      {
        "name": "name",
        "type": "nbtTagName"
      },
      {
        "name": "value",
        "type": [
          "nbtSwitch",
          {
            "type": "type"
          }
        ]
      }
    ]
  ]
}
```

The tests run ok disabling the validation

Comment (rom1504) [reply bench/compiled_nbt.js:12]: is some type not added as native? how come it fails validation?

File: nbt.json (hunk 4 above)
Comment (rom1504) [reply nbt.json:158]: I think it makes sense to have them here

File: bench/compiled_nbt.js (hunk 5 above)
Comment (extremeheat) [reply bench/compiled_nbt.js]: Ah yeah looks like missing native was the cause, should be fixed now

**Comment by extremeheat** [comment]
Validator issue is fixed, just need to check over the protocol.json now. But anyway I think this can be merged before that, would just need to be checking before working on mineflayer

**Comment by rom1504** [comment]
thanks

**Pull request merged**